### PR TITLE
[Messenger] Uses Symfony Serializer by default for envelope items

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28164
| License       | MIT
| Doc PR        | ø

The original approach was to use `serialize`/`unserialize` for envelope items. It turns out it has limitations (see #28247) and reduces the compatibility with 3rd party systems (see #28164). This pull-request changes the existing mechanism by using Symfony Serializer by default. 

It keeps the `serialize`/`unserialize` mechanism to allow users of the experimental component to keep using it in 4.2 and allow to have a non-disruptive upgrade to 4.2.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 1]
```diff
@@ -22,12 +22,14 @@ class Serializer implements DecoderInterface, EncoderInterface
     private $serializer;
     private $format;
     private $context;
+    private $envelopeWithSerialize;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: encodeUsingPhpSerialize? (see other comment about "decode")

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 2]
```diff
@@ -67,14 +69,49 @@ public function encode(Envelope $envelope): array
...
 
         return array(
             'body' => $this->serializer->serialize($envelope->getMessage(), $this->format, $context),
             'headers' => $headers,
         );
     }
+
+    private function decodeEnvelopeItems($encodedEnvelope)
+    {
+        if ($this->envelopeWithSerialize) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: should this flag be set? isn't `isset($encodedEnvelope['headers']['X-Message-Envelope-Items'])` enough?

Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: You are right, it would be better 👌

Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php]: Updated 👍

**Comment by ogizanagi** [comment]
Rational in original PR was that envelop items are a thing specific to Symfony Messenger component to configure transport & middleware and would not be used by third party systems. So there is no interop required (or that would means 3rd party systems would have to implement features the same way Symfony Messenger does. Not likely to happen IMHO).

Mentioned use-case in #28270 is about adding headers to the transport layer. Hence this is the transport responsibility. Using directly envelop items for this and relying on the fact those are serialized in `X-Message-Envelope-Items` looks wrong to me. Instead, I'd say an envelope item should allow configuring headers and our Messenger Serializer uses it to add asked headers.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
👍  for using plain JSON via Symfony Serializer (we are moving away from PHP serializer anyway).
👎 for keeping BC with 4.1. The component is marked as experimental for this very reason: allow us to break BC.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sroze.

**Comment by javiereguiluz** [comment]
I've created https://github.com/symfony/symfony-docs/issues/10276 to document this new feature. Please, don't forget to create a doc issue for every new feature.

Also, we have lots of [pending issues related to Messenger](https://github.com/symfony/symfony-docs/labels/Messenger) in the Symfony Docs. We really need your help 🙏  Thanks!

**Pull request merged**